stake-pool/program/tests/withdraw.rs

```diff
@@ -151,3 +196,736 @@ async fn test_stake_pool_withdraw() {
+
+    let tokens_to_burn = pool_tokens / 4;
+    let lamports_to_withdraw = tokens_to_burn; // For now math is 1:1
+
+    // Delegate tokens for burning
+    delegate_tokens(
+        &mut banks_client,
+        &payer,
+        &recent_blockhash,
+        &user_pool_account,
+        &user,
+        &stake_pool_accounts.withdraw_authority,
+        tokens_to_burn,
+    )
+    .await;
+
+    // Create stake account to withdraw to
+    let user_stake_recipient = Keypair::new();
+
+    let new_authority = Pubkey::new_unique();
+
+    let transaction_error = stake_pool_accounts
+        .withdraw_stake(
+            &mut banks_client,
+            &payer,
+            &recent_blockhash,
+            &user_stake_recipient.pubkey(),
+            &user_pool_account,
+            &validator_stake_account.stake_account,
+            &new_authority,
+            lamports_to_withdraw,
+        )
+        .await
+        .err()
+        .unwrap();
+
+    match transaction_error {
+        TransportError::TransactionError(TransactionError::InstructionError(
+            _,
+            InstructionError::Custom(error_index),
+        )) => {
+            let program_error = error::StakePoolError::ValidatorNotFound as u32;
+            assert_eq!(error_index, program_error);
+        }
+        _ => panic!("Wrong error occurs while try to do withdraw from unknown validator"),
+    }
+}
+
+#[tokio::test]
+async fn test_stake_pool_double_withdraw() {
+    let (
+        mut banks_client,
+        payer,
+        recent_blockhash,
+        stake_pool_accounts,
+        validator_stake_account,
+        deposit_info,
+        _,
+        lamports_to_withdraw,
+    ) = setup().await;
+
+    // Create stake account to withdraw to
+    let user_stake_recipient = Keypair::new();
+    create_blank_stake_account(
+        &mut banks_client,
+        &payer,
+        &recent_blockhash,
+        &user_stake_recipient,
+    )
+    .await;
+
+    let new_authority = Pubkey::new_unique();
+    stake_pool_accounts
+        .withdraw_stake(
+            &mut banks_client,
+            &payer,
+            &recent_blockhash,
+            &user_stake_recipient.pubkey(),
+            &deposit_info.user_pool_account,
+            &validator_stake_account.stake_account,
+            &new_authority,
+            lamports_to_withdraw,
+        )
+        .await
+        .unwrap();
+
+    let latest_blockhash = banks_client.get_recent_blockhash().await.unwrap();
+
+    let transaction_error = stake_pool_accounts
+        .withdraw_stake(
+            &mut banks_client,
+            &payer,
+            &latest_blockhash,
+            &user_stake_recipient.pubkey(),
+            &deposit_info.user_pool_account,
+            &validator_stake_account.stake_account,
+            &new_authority,
+            lamports_to_withdraw,
+        )
+        .await
+        .err()
+        .unwrap();
+
+    match transaction_error {
```

Comment:
I might be missing something, so let me know if I have this right -- we're trying to withdraw 1/4 from the same stake twice, so it should normally succeed, but we're missing another call to delegate the pool tokens.  Is this trying to test the token approval code?  On a separate note, why does this fail with `InvalidAccountData`?

Reply:
Actually this is a test of double withdrawal to the same stake account and the error comes from the stake program itself (because stake account on the second withdraw is not a new one and already initialized).

Reply:
Oh nice! Right, that makes sense, thanks for clarifying